Lib/idlelib/idle_test/run_coverage.py

```diff
@@ -0,0 +1,179 @@
+        if sys.platform == 'win32' else
+        venv_path.joinpath('bin', 'python')
+    )
+    built_python_path = sources_root_path.joinpath(
+        'python.bat' if sys.platform == 'win32' else
+        'python.exe' if sys.platform == 'darwin' else
+        'python'
+    )
+    if not venv_path.is_dir():
+        subprocess.run([built_python_path, '-m', 'venv', venv_path])
+        subprocess.run([venv_python_path, '-m', 'pip', 'install', 'coverage'])
+    return venv_path, venv_python_path
+
+
+@contextlib.contextmanager
+def coveragerc_replacement():
+    coveragerc_path = pathlib.Path('.coveragerc')
+    orig_coverage_rc = None
+    if coveragerc_path.is_file():
+        with coveragerc_path.open('r', encoding='utf-8') as f:
+            orig_coverage_rc = f.read()
+    try:
+        with coveragerc_path.open('w', encoding='utf-8') as f:
+            f.write(textwrap.dedent('''\
+                [run]
+                branch = True
+                cover_pylib = True
+
+                [report]
+                # Regexes for lines to exclude from consideration
+                exclude_lines =
+                    # Don't complain if non-runnable code isn't run:
+                    if 0:
+                    if False:
+                    if __name__ == .__main__.:
+
+                    .*# htest #
+                    if not _utest:
+                    if _htest:
+                show_missing = True
+                '''))
+        yield
+    finally:
+        if orig_coverage_rc:
+            with coveragerc_path.open('w', encoding='utf-8') as f:
+                f.write(orig_coverage_rc)
+        else:
+            coveragerc_path.unlink()
+
+
+def run_tests_with_coverage(*, module_name, venv_python_path):
+    if module_name == 'all':
+        subprocess.run([
+            venv_python_path, '-m', 'coverage', 'run',
+            '--source=idlelib',
+            '-m', 'test', '-ugui', 'test_idle',
+        ])
+    else:
+        subprocess.run([
+            venv_python_path, '-m', 'coverage', 'run',
+            f'--source=idlelib.{module_name}',
+            f'./Lib/idlelib/idle_test/test_{module_name}.py',
+        ])
+        if module_name in ['pyshell', 'run']:
+            # also run the tests in test_warning.py
+            subprocess.run([
+                venv_python_path, '-m', 'coverage', 'run', '-a',
+                f'--source=idlelib.{module_name}',
+                './Lib/idlelib/idle_test/test_warning.py',
+            ])
+
+
+def generate_coverage_report(*, venv_python_path, no_html):
+    subprocess.run([venv_python_path, '-m', 'coverage', 'report'])
+    if not no_html:
+        subprocess.run([venv_python_path, '-m', 'coverage', 'html'])
+
+
+def display_coverage_report(*, module_name):
+    html_file = (
+        'htmlcov/index.html'
+        if module_name == 'all' else
+        f'htmlcov/Lib_idlelib_{module_name}_py.html'
+    )
+    webbrowser.open(html_file)
```

Comment:
On Windows, this only works if `webbrowser` is given an absolute path.
On macOS, it does not work at all.[^1]

Perhaps it would be better to just print the path and let the user open the report manually. OTOH, we could deny this feature on macOS.

Also, we should use `index.html` always; `coverage` updates it correctly depending on what `module_name` is. The computed paths in this PR are broken, since `coverage` uses filenames with has prefixes.

```suggestion
    cwd = pathlib.Path(os.getcwd())
    path = cwd.joinpath('htmlcov', 'index.html')
    if sys.platform == 'Darwin':
        print(f'Coverage results published to {path}')
    else:
        webbrowser.open(path)
```

[^1]: `webbrowser` does not even work using the CLI on macOS